{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Genotypic recalibration with machine learning: usage example\n",
    "\n",
    "This notebook illustrates the use of machine learning methods and associated code described in the parent paper."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "import sys\n",
    "sys.path.insert(0, '../python')\n",
    "\n",
    "from preprocessing import VCF, load_suffixes, prepare_input\n",
    "from recalibrator import Recalibrator"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Training & saving a model\n",
    "\n",
    "Training is performed on a dataset consisting of VCF files resulting from performing variant calling with GATK on reads from a family trio, along with a 'synthetic abortus' that contains a mixture of the mother's and child's reads. The code to read and process the dataset relies on a specific directory structure. The `VCF` class and the `prepare input` method are all that's needed to read a VCF and convert it into an array that can be input into a model. Once trained, a recalibrator model can be serialized and saved."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "trios = [\"ajt\", \"chd\", \"corpas\", \"yri\"]\n",
    "\n",
    "# Pre-processing. Uncomment during first run of the script, then\n",
    "# comment to avoid re-computing\n",
    "\n",
    "# for trio in trios:\n",
    "#     data_dir = '../data/' + trio + '/'\n",
    "#     df = load_suffixes(data_dir)\n",
    "#     df.to_csv(trio + '.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Construct training dataset by concatenating rows from all the synthetic abortus trios"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_train = pd.DataFrame()\n",
    "    \n",
    "for train in trios:\n",
    "    df_train = df_train.append(pd.read_csv(train + '.csv'))\n",
    "    \n",
    "df_train = df_train[::100] # Train on subset of input rows"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* Split dataset into input and target parameters\n",
    "* Initialize a recalibrator model and train it on our data\n",
    "* Save the serialized model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Training logistic regression\n",
      "Training XGB\n",
      "[0]\tvalidation_0-merror:0.056753\n",
      "Will train until validation_0-merror hasn't improved in 20 rounds.\n",
      "[1]\tvalidation_0-merror:0.054217\n",
      "[2]\tvalidation_0-merror:0.053266\n",
      "[3]\tvalidation_0-merror:0.051997\n",
      "[4]\tvalidation_0-merror:0.05168\n",
      "[5]\tvalidation_0-merror:0.050729\n",
      "[6]\tvalidation_0-merror:0.049144\n",
      "[7]\tvalidation_0-merror:0.048193\n",
      "[8]\tvalidation_0-merror:0.049461\n",
      "[9]\tvalidation_0-merror:0.047559\n",
      "[10]\tvalidation_0-merror:0.045656\n",
      "[11]\tvalidation_0-merror:0.044705\n",
      "[12]\tvalidation_0-merror:0.044071\n",
      "[13]\tvalidation_0-merror:0.044071\n",
      "[14]\tvalidation_0-merror:0.044071\n",
      "[15]\tvalidation_0-merror:0.043754\n",
      "[16]\tvalidation_0-merror:0.04312\n",
      "[17]\tvalidation_0-merror:0.042803\n",
      "[18]\tvalidation_0-merror:0.043437\n",
      "[19]\tvalidation_0-merror:0.042486\n",
      "[20]\tvalidation_0-merror:0.042486\n",
      "[21]\tvalidation_0-merror:0.042169\n",
      "[22]\tvalidation_0-merror:0.042803\n",
      "[23]\tvalidation_0-merror:0.04312\n",
      "[24]\tvalidation_0-merror:0.041218\n",
      "[25]\tvalidation_0-merror:0.040583\n",
      "[26]\tvalidation_0-merror:0.0409\n",
      "[27]\tvalidation_0-merror:0.0409\n",
      "[28]\tvalidation_0-merror:0.039632\n",
      "[29]\tvalidation_0-merror:0.038681\n",
      "[30]\tvalidation_0-merror:0.038681\n",
      "[31]\tvalidation_0-merror:0.039315\n",
      "[32]\tvalidation_0-merror:0.03773\n",
      "[33]\tvalidation_0-merror:0.038364\n",
      "[34]\tvalidation_0-merror:0.038681\n",
      "[35]\tvalidation_0-merror:0.038364\n",
      "[36]\tvalidation_0-merror:0.038047\n",
      "[37]\tvalidation_0-merror:0.037413\n",
      "[38]\tvalidation_0-merror:0.036462\n",
      "[39]\tvalidation_0-merror:0.036779\n",
      "[40]\tvalidation_0-merror:0.036145\n",
      "[41]\tvalidation_0-merror:0.036462\n",
      "[42]\tvalidation_0-merror:0.03551\n",
      "[43]\tvalidation_0-merror:0.036145\n",
      "[44]\tvalidation_0-merror:0.036462\n",
      "[45]\tvalidation_0-merror:0.036462\n",
      "[46]\tvalidation_0-merror:0.036145\n",
      "[47]\tvalidation_0-merror:0.036145\n",
      "[48]\tvalidation_0-merror:0.036145\n",
      "[49]\tvalidation_0-merror:0.035828\n",
      "[50]\tvalidation_0-merror:0.036462\n",
      "[51]\tvalidation_0-merror:0.036462\n",
      "[52]\tvalidation_0-merror:0.035828\n",
      "[53]\tvalidation_0-merror:0.036462\n",
      "[54]\tvalidation_0-merror:0.03551\n",
      "[55]\tvalidation_0-merror:0.03551\n",
      "[56]\tvalidation_0-merror:0.03551\n",
      "[57]\tvalidation_0-merror:0.03551\n",
      "[58]\tvalidation_0-merror:0.036145\n",
      "[59]\tvalidation_0-merror:0.03551\n",
      "[60]\tvalidation_0-merror:0.03551\n",
      "[61]\tvalidation_0-merror:0.035828\n",
      "[62]\tvalidation_0-merror:0.03551\n",
      "Stopping. Best iteration:\n",
      "[42]\tvalidation_0-merror:0.03551\n",
      "\n",
      "Training meta-classifier\n"
     ]
    }
   ],
   "source": [
    "X_train = prepare_input(df_train, target_cols=['justchild^GT'])\n",
    "y_train = df_train['justchild^GT'].values\n",
    "\n",
    "r = Recalibrator()\n",
    "r.train(X_train, y_train)\n",
    "r.save(\"model.pickle\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Loading a model & recalibrating"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* Instantiate a recalibrator object and load a saved model\n",
    "* Instantiate a VCF object from a VCF file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>ID</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>QUAL</th>\n",
       "      <th>FILTER</th>\n",
       "      <th>INFO</th>\n",
       "      <th>FORMAT</th>\n",
       "      <th>abortus</th>\n",
       "      <th>father</th>\n",
       "      <th>mother</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>69270</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>91.17</td>\n",
       "      <td>.</td>\n",
       "      <td>AC=2;AF=1.00;AN=2;DP=4;ExcessHet=3.0103;FS=0.0...</td>\n",
       "      <td>GT:AD:DP:GQ:PL</td>\n",
       "      <td>./.:0,0:0:.:0,0,0</td>\n",
       "      <td>1/1:0,4:4:12:117,12,0</td>\n",
       "      <td>./.:0,0:0:.:0,0,0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>69511</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>3514.90</td>\n",
       "      <td>.</td>\n",
       "      <td>AC=2;AF=1.00;AN=2;DP=128;ExcessHet=3.0103;FS=0...</td>\n",
       "      <td>GT:AD:DP:GQ:PL</td>\n",
       "      <td>./.:0,0:0:.:0,0,0</td>\n",
       "      <td>1/1:0,128:128:99:3541,383,0</td>\n",
       "      <td>./.:0,0:0:.:0,0,0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>183238</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>20.18</td>\n",
       "      <td>.</td>\n",
       "      <td>AC=1;AF=0.250;AN=6;BaseQRankSum=-1.282e+00;Cli...</td>\n",
       "      <td>GT:AD:DP:GQ:PL</td>\n",
       "      <td>0/0:9,0:9:27:0,27,231</td>\n",
       "      <td>0/0:14,0:14:39:0,39,585</td>\n",
       "      <td>0/1:3,3:6:51:51,0,81</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>187485</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>244.19</td>\n",
       "      <td>.</td>\n",
       "      <td>AC=5;AF=0.750;AN=6;BaseQRankSum=-9.670e-01;Cli...</td>\n",
       "      <td>GT:AD:DP:GQ:PGT:PID:PL</td>\n",
       "      <td>1/1:0,2:2:6:.:.:49,6,0</td>\n",
       "      <td>0/1:1,2:3:19:.:.:40,0,19</td>\n",
       "      <td>1/1:0,5:5:15:1|1:187485_G_A:185,15,0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>942451</td>\n",
       "      <td>.</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>980.92</td>\n",
       "      <td>.</td>\n",
       "      <td>AC=6;AF=1.00;AN=6;DP=35;ExcessHet=3.0103;FS=0....</td>\n",
       "      <td>GT:AD:DP:GQ:PL</td>\n",
       "      <td>1/1:0,20:20:60:561,60,0</td>\n",
       "      <td>1/1:0,8:8:24:242,24,0</td>\n",
       "      <td>1/1:0,7:7:21:204,21,0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  #CHROM     POS ID REF ALT     QUAL FILTER  \\\n",
       "0   chr1   69270  .   A   G    91.17      .   \n",
       "1   chr1   69511  .   A   G  3514.90      .   \n",
       "2   chr1  183238  .   G   C    20.18      .   \n",
       "3   chr1  187485  .   G   A   244.19      .   \n",
       "4   chr1  942451  .   T   C   980.92      .   \n",
       "\n",
       "                                                INFO                  FORMAT  \\\n",
       "0  AC=2;AF=1.00;AN=2;DP=4;ExcessHet=3.0103;FS=0.0...          GT:AD:DP:GQ:PL   \n",
       "1  AC=2;AF=1.00;AN=2;DP=128;ExcessHet=3.0103;FS=0...          GT:AD:DP:GQ:PL   \n",
       "2  AC=1;AF=0.250;AN=6;BaseQRankSum=-1.282e+00;Cli...          GT:AD:DP:GQ:PL   \n",
       "3  AC=5;AF=0.750;AN=6;BaseQRankSum=-9.670e-01;Cli...  GT:AD:DP:GQ:PGT:PID:PL   \n",
       "4  AC=6;AF=1.00;AN=6;DP=35;ExcessHet=3.0103;FS=0....          GT:AD:DP:GQ:PL   \n",
       "\n",
       "                   abortus                       father  \\\n",
       "0        ./.:0,0:0:.:0,0,0        1/1:0,4:4:12:117,12,0   \n",
       "1        ./.:0,0:0:.:0,0,0  1/1:0,128:128:99:3541,383,0   \n",
       "2    0/0:9,0:9:27:0,27,231      0/0:14,0:14:39:0,39,585   \n",
       "3   1/1:0,2:2:6:.:.:49,6,0     0/1:1,2:3:19:.:.:40,0,19   \n",
       "4  1/1:0,20:20:60:561,60,0        1/1:0,8:8:24:242,24,0   \n",
       "\n",
       "                                 mother  \n",
       "0                     ./.:0,0:0:.:0,0,0  \n",
       "1                     ./.:0,0:0:.:0,0,0  \n",
       "2                  0/1:3,3:6:51:51,0,81  \n",
       "3  1/1:0,5:5:15:1|1:187485_G_A:185,15,0  \n",
       "4                 1/1:0,7:7:21:204,21,0  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "r = Recalibrator()\n",
    "r.load(\"model.pickle\")\n",
    "\n",
    "abortus = VCF(\"../data/ajt/abortus.frac0.5.seed151_trio.vcf\")\n",
    "abortus.df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Estimated contamination: 0.5066857431635673\n"
     ]
    }
   ],
   "source": [
    "# VCF class automatically infers the contamination fraction\n",
    "abortus.process(\"abortus\", \"mother\", \"father\")\n",
    "print(\"Estimated contamination: {}\".format(abortus.estimated_contamination))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>AC</th>\n",
       "      <th>AF</th>\n",
       "      <th>abortus^GT</th>\n",
       "      <th>abortus^GQ</th>\n",
       "      <th>abortus^DP</th>\n",
       "      <th>mother^GT</th>\n",
       "      <th>...</th>\n",
       "      <th>abortus^GT^0^1H</th>\n",
       "      <th>abortus^GT^1^1H</th>\n",
       "      <th>abortus^GT^2^1H</th>\n",
       "      <th>mother^GT^0^1H</th>\n",
       "      <th>mother^GT^1^1H</th>\n",
       "      <th>mother^GT^2^1H</th>\n",
       "      <th>father^GT^0^1H</th>\n",
       "      <th>father^GT^1^1H</th>\n",
       "      <th>father^GT^2^1H</th>\n",
       "      <th>contamination</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>183238</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0.25</td>\n",
       "      <td>0</td>\n",
       "      <td>27</td>\n",
       "      <td>9</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>0.506686</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>187485</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>0.75</td>\n",
       "      <td>2</td>\n",
       "      <td>6</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>0.506686</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>942451</td>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "      <td>6</td>\n",
       "      <td>1.00</td>\n",
       "      <td>2</td>\n",
       "      <td>60</td>\n",
       "      <td>20</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>0.506686</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>942934</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>0.50</td>\n",
       "      <td>1</td>\n",
       "      <td>99</td>\n",
       "      <td>184</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>0.506686</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>943937</td>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "      <td>0.25</td>\n",
       "      <td>1</td>\n",
       "      <td>99</td>\n",
       "      <td>88</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>0.506686</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 40 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  #CHROM     POS  REF  ALT  AC    AF  abortus^GT  abortus^GQ  abortus^DP  \\\n",
       "0   chr1  183238    3    2   1  0.25           0          27           9   \n",
       "1   chr1  187485    3    1   5  0.75           2           6           2   \n",
       "2   chr1  942451    4    2   6  1.00           2          60          20   \n",
       "3   chr1  942934    3    2   3  0.50           1          99         184   \n",
       "4   chr1  943937    2    4   2  0.25           1          99          88   \n",
       "\n",
       "   mother^GT      ...        abortus^GT^0^1H  abortus^GT^1^1H  \\\n",
       "0          1      ...                   True            False   \n",
       "1          2      ...                  False            False   \n",
       "2          2      ...                  False            False   \n",
       "3          0      ...                  False             True   \n",
       "4          1      ...                  False             True   \n",
       "\n",
       "   abortus^GT^2^1H  mother^GT^0^1H  mother^GT^1^1H  mother^GT^2^1H  \\\n",
       "0            False           False            True           False   \n",
       "1             True           False           False            True   \n",
       "2             True           False           False            True   \n",
       "3            False            True           False           False   \n",
       "4            False           False            True           False   \n",
       "\n",
       "   father^GT^0^1H  father^GT^1^1H  father^GT^2^1H  contamination  \n",
       "0            True           False           False       0.506686  \n",
       "1           False            True           False       0.506686  \n",
       "2           False           False            True       0.506686  \n",
       "3           False           False            True       0.506686  \n",
       "4            True           False           False       0.506686  \n",
       "\n",
       "[5 rows x 40 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "abortus.df_processed.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Predicted labels\n",
    "preds_lr = r.predict_lr(abortus.prepare_input())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "abortus.save_predictions(preds_lr, filename=\"recalibrated_lr.vcf\", sample=\"abortus\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Recalibrating with confidence intervals\n",
    "\n",
    "We use the `VCF` class's inbuilt method to process the VCF and extract the fields required by `confidence_intervals`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "abortus = VCF(\"../data/ajt/abortus.frac0.5.seed151_trio.vcf\")\n",
    "abortus.process(\"abortus\", \"mother\", \"father\")\n",
    "preds_ci = r.model_ci.predict(abortus.prepare_input())\n",
    "\n",
    "# abortus.save_predictions(preds_ci, filename=\"recalibrated_ci.vcf\", sample=\"abortus\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
